core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java

```diff
@@ -0,0 +1,234 @@
+                if (fp.getOriginal() != null) {
+                    w.println("  <original>");
+                    w.print("    <name>");
+                    w.print(Util.xmlEscape(fp.getOriginal().getName()));
+                    w.println("</name>");
+                    w.print("    <number>");
+                    w.print(fp.getOriginal().getNumber());
+                    w.println("</number>");
+                    w.println("  </original>");
+                }
+                // TODO(oleg_nenashev): Consider renaming the field: https://issues.jenkins-ci.org/browse/JENKINS-25808
+                w.print("  <md5sum>");
+                w.print(fp.getHashString());
+                w.println("</md5sum>");
+                w.print("  <fileName>");
+                w.print(Util.xmlEscape(fp.getFileName()));
+                w.println("</fileName>");
+                w.println("  <usages>");
+                for (Map.Entry<String, Fingerprint.RangeSet> e : fp.getUsages().entrySet()) {
+                    w.println("    <entry>");
+                    w.print("      <string>");
+                    w.print(Util.xmlEscape(e.getKey()));
+                    w.println("</string>");
+                    w.print("      <ranges>");
+                    w.print(serialize(e.getValue()));
+                    w.println("</ranges>");
+                    w.println("    </entry>");
+                }
+                w.println("  </usages>");
+                w.println("  <facets/>");
+                w.print("</fingerprint>");
+                w.flush();
+                afw.commit();
+            } finally {
+                afw.abort();
+            }
+        } else {
+            // Slower fallback that can persist facets.
+            getConfigFile(file).write(fp);
+        }
+    }
+
+    /**
+     * Deletes the Fingerprint with the given unique ID.
+     */
+    public void delete(String id) {
+        File file = getFingerprintFile(id);
+        file.delete();
+
+        File inner = new File(Jenkins.get().getRootDir(), "fingerprints/"+id.substring(0,2)+"/"+id.substring(2,4));
+        String[] inner_files = inner.list();
+        if (inner_files!=null && inner_files.length==0) inner.delete();
+
+        File outer = new File(Jenkins.get().getRootDir(), "fingerprints/"+id.substring(0,2));
+        String[] outer_files = outer.list();
+        if (outer_files!=null && outer_files.length==0) outer.delete();
+    }
+
+    /**
+     * The file we save our configuration.
+     */
+    private static @NonNull XmlFile getConfigFile(@NonNull File file) {
+        return new XmlFile(Fingerprint.getXStream(), file);
+    }
+
+    /**
+     * Determines the file name from unique id (md5sum).
+     */
+    private static @NonNull File getFingerprintFile(@NonNull String id) {
+        return new File( Jenkins.get().getRootDir(),
+                "fingerprints/"+id.substring(0,2)+'/'+id.substring(2,4)+'/'+id.substring(4)+".xml");
```

Comment:
Same comment here: non-portable path concatenation

Reply:
@afalko I agree with you, but interestingly, in the Jenkins core codebase, I can mostly find non-portable path concatenation being used.
CC @oleg-nenashev some insights would helpful :)